File: web/src/app/chat/services/streamingModels.ts
import { OnyxDocument } from "@/lib/search/interfaces";

// Base interface for all streaming objects
interface BaseObj {
  type: string;
}

export enum PacketType {
  MESSAGE_START = "message_start",
  MESSAGE_DELTA = "message_delta",
  MESSAGE_END = "message_end",

  STOP = "stop",
  SECTION_END = "section_end",

  // Specific tool packets
  SEARCH_TOOL_START = "internal_search_tool_start",
  SEARCH_TOOL_DELTA = "internal_search_tool_delta",
  IMAGE_GENERATION_TOOL_START = "image_generation_tool_start",
  IMAGE_GENERATION_TOOL_DELTA = "image_generation_tool_delta",

  // Custom tool packets
  CUSTOM_TOOL_START = "custom_tool_start",
  CUSTOM_TOOL_DELTA = "custom_tool_delta",

  // Reasoning packets
  REASONING_START = "reasoning_start",
  REASONING_DELTA = "reasoning_delta",
  REASONING_END = "reasoning_end",

  CITATION_START = "citation_start",
  CITATION_DELTA = "citation_delta",
  CITATION_END = "citation_end",
}

// Basic Message Packets
export interface MessageStart extends BaseObj {
  id: string;
  type: "message_start";
  content: string;

  final_documents: OnyxDocument[] | null;
}

export interface MessageDelta extends BaseObj {
  content: string;
  type: "message_delta";
}

export interface MessageEnd extends BaseObj {
  type: "message_end";
}

// Control Packets
export interface Stop extends BaseObj {
  type: "stop";
}

export interface SectionEnd extends BaseObj {
  type: "section_end";
}

// Specific tool packets
export interface SearchToolStart extends BaseObj {
  type: "internal_search_tool_start";
  is_internet_search?: boolean;
}

export interface SearchToolDelta extends BaseObj {
  type: "internal_search_tool_delta";
  queries: string[] | null;
  documents: OnyxDocument[] | null;
}

interface GeneratedImage {
  file_id: string;
  url: string;
  revised_prompt: string;
}

export interface ImageGenerationToolStart extends BaseObj {
  type: "image_generation_tool_start";
}

export interface ImageGenerationToolDelta extends BaseObj {
  type: "image_generation_tool_delta";
  images: GeneratedImage[];
}

// Custom Tool Packets
export interface CustomToolStart extends BaseObj {
  type: "custom_tool_start";
  tool_name: string;
}

export interface CustomToolDelta extends BaseObj {
  type: "custom_tool_delta";
  tool_name: string;
  response_type: string;
  data?: any;
  file_ids?: string[] | null;
}

// Reasoning Packets
export interface ReasoningStart extends BaseObj {
  type: "reasoning_start";
}

export interface ReasoningDelta extends BaseObj {
  type: "reasoning_delta";
  reasoning: string;
}

// Citation Packets
export interface StreamingCitation {
  citation_num: number;
  document_id: string;
}

export interface CitationStart extends BaseObj {
  type: "citation_start";
}

export interface CitationDelta extends BaseObj {
  type: "citation_delta";
  citations: StreamingCitation[];
}

export type ChatObj = MessageStart | MessageDelta | MessageEnd;

export type StopObj = Stop;

export type SectionEndObj = SectionEnd;

// Specific tool objects
export type SearchToolObj = SearchToolStart | SearchToolDelta | SectionEnd;
export type ImageGenerationToolObj =
  | ImageGenerationToolStart
  | ImageGenerationToolDelta
  | SectionEnd;
export type CustomToolObj = CustomToolStart | CustomToolDelta | SectionEnd;
export type NewToolObj = SearchToolObj | ImageGenerationToolObj | CustomToolObj;

export type ReasoningObj = ReasoningStart | ReasoningDelta | SectionEnd;

export type CitationObj = CitationStart | CitationDelta | SectionEnd;

// Union type for all possible streaming objects
export type ObjTypes =
  | ChatObj
  | NewToolObj
  | ReasoningObj
  | StopObj
  | SectionEndObj
  | CitationObj;

// Packet wrapper for streaming objects
export interface Packet {
  ind: number;
  obj: ObjTypes;
}

export interface ChatPacket {
  ind: number;
  obj: ChatObj;
}

export interface StopPacket {
  ind: number;
  obj: StopObj;
}

export interface CitationPacket {
  ind: number;
  obj: CitationObj;
}

// New specific tool packet types
export interface SearchToolPacket {
  ind: number;
  obj: SearchToolObj;
}

export interface ImageGenerationToolPacket {
  ind: number;
  obj: ImageGenerationToolObj;
}

export interface CustomToolPacket {
  ind: number;
  obj: CustomToolObj;
}

export interface ReasoningPacket {
  ind: number;
  obj: ReasoningObj;
}

export interface SectionEndPacket {
  ind: number;
  obj: SectionEndObj;
}

File: backend/onyx/server/query_and_chat/streaming_models.py
from collections import OrderedDict
from collections.abc import Mapping
from enum import Enum
from typing import Annotated
from typing import Literal
from typing import Union

from pydantic import BaseModel
from pydantic import Field

from onyx.agents.agent_search.dr.models import GeneratedImage
from onyx.context.search.models import SavedSearchDoc


class BaseObj(BaseModel):
    type: str = ""


"""Basic Message Packets"""


class MessageStart(BaseObj):
    type: Literal["message_start"] = "message_start"

    # Merged set of all documents considered
    final_documents: list[SavedSearchDoc] | None

    content: str


class MessageDelta(BaseObj):
    content: str
    type: Literal["message_delta"] = "message_delta"


"""Control Packets"""


class OverallStop(BaseObj):
    type: Literal["stop"] = "stop"


class SectionEnd(BaseObj):
    type: Literal["section_end"] = "section_end"


"""Tool Packets"""


class SearchToolStart(BaseObj):
    type: Literal["internal_search_tool_start"] = "internal_search_tool_start"

    is_internet_search: bool = False


class SearchToolDelta(BaseObj):
    type: Literal["internal_search_tool_delta"] = "internal_search_tool_delta"

    queries: list[str] | None = None
    documents: list[SavedSearchDoc] | None = None


class ImageGenerationToolStart(BaseObj):
    type: Literal["image_generation_tool_start"] = "image_generation_tool_start"


class ImageGenerationToolDelta(BaseObj):
    type: Literal["image_generation_tool_delta"] = "image_generation_tool_delta"

    images: list[GeneratedImage]


class CustomToolStart(BaseObj):
    type: Literal["custom_tool_start"] = "custom_tool_start"

    tool_name: str


class CustomToolDelta(BaseObj):
    type: Literal["custom_tool_delta"] = "custom_tool_delta"

    tool_name: str
    response_type: str
    # For non-file responses
    data: dict | list | str | int | float | bool | None = None
    # For file-based responses like image/csv
    file_ids: list[str] | None = None


"""Reasoning Packets"""


class ReasoningStart(BaseObj):
    type: Literal["reasoning_start"] = "reasoning_start"


class ReasoningDelta(BaseObj):
    type: Literal["reasoning_delta"] = "reasoning_delta"

    reasoning: str


"""Citation Packets"""


class CitationStart(BaseObj):
    type: Literal["citation_start"] = "citation_start"


class SubQuestionIdentifier(BaseModel):
    """None represents references to objects in the original flow. To our understanding,
    these will not be None in the packets returned from agent search.
    """

    level: int | None = None
    level_question_num: int | None = None

    @staticmethod
    def make_dict_by_level(
        original_dict: Mapping[tuple[int, int], "SubQuestionIdentifier"],
    ) -> dict[int, list["SubQuestionIdentifier"]]:
        """returns a dict of level to object list (sorted by level_question_num)
        Ordering is asc for readability.
        """

        # organize by level, then sort ascending by question_index
        level_dict: dict[int, list[SubQuestionIdentifier]] = {}

        # group by level
        for k, obj in original_dict.items():
            level = k[0]
            if level not in level_dict:
                level_dict[level] = []
            level_dict[level].append(obj)

        # for each level, sort the group
        for k2, value2 in level_dict.items():
            # we need to handle the none case due to SubQuestionIdentifier typing
            # level_question_num as int | None, even though it should never be None here.
            level_dict[k2] = sorted(
                value2,
                key=lambda x: (x.level_question_num is None, x.level_question_num),
            )

        # sort by level
        sorted_dict = OrderedDict(sorted(level_dict.items()))
        return sorted_dict


class CitationInfo(SubQuestionIdentifier):
    citation_num: int
    document_id: str


class CitationDelta(BaseObj):
    type: Literal["citation_delta"] = "citation_delta"

    citations: list[CitationInfo] | None = None


"""Packet"""

# Discriminated union of all possible packet object types
PacketObj = Annotated[
    Union[
        MessageStart,
        MessageDelta,
        OverallStop,
        SectionEnd,
        SearchToolStart,
        SearchToolDelta,
        ImageGenerationToolStart,
        ImageGenerationToolDelta,
        CustomToolStart,
        CustomToolDelta,
        ReasoningStart,
        ReasoningDelta,
        CitationStart,
        CitationDelta,
    ],
    Field(discriminator="type"),
]


class Packet(BaseModel):
    ind: int
    obj: PacketObj


class EndStepPacketList(BaseModel):
    end_step_nr: int
    packet_list: list[Packet]


class StreamingType(Enum):
    MESSAGE_START = "message_start"
    MESSAGE_DELTA = "message_delta"
    INTERNAL_SEARCH_TOOL_START = "internal_search_tool_start"
    INTERNAL_SEARCH_TOOL_DELTA = "internal_search_tool_delta"
    IMAGE_GENERATION_TOOL_START = "image_generation_tool_start"
    IMAGE_GENERATION_TOOL_DELTA = "image_generation_tool_delta"
    REASONING_START = "reasoning_start"
    REASONING_DELTA = "reasoning_delta"
    CITATION_START = "citation_start"
    CITATION_DELTA = "citation_delta"
    CUSTOM_TOOL_START = "custom_tool_start"
    CUSTOM_TOOL_DELTA = "custom_tool_delta"
